client/alloc_endpoint.go

```diff
@@ -100,3 +116,186 @@ func (a *Allocations) Stats(args *cstructs.AllocStatsRequest, reply *cstructs.Al
 	reply.Stats = stats
 	return nil
 }
+
+func (a *Allocations) exec(conn io.ReadWriteCloser) {
+	defer metrics.MeasureSince([]string{"client", "allocations", "exec"}, time.Now())
+	defer conn.Close()
+
+	// Decode the arguments
+	var req cstructs.AllocExecRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	aclObj, token, err := a.c.resolveTokenAndACL(req.QueryOptions.AuthToken)
+	{
+		// log access
+		execID := uuid.Generate()
+		tokenName, tokenID := "", ""
+		if token != nil {
+			tokenName, tokenID = token.Name, token.AccessorID
+		}
+
+		a.c.logger.Info("task exec session starting",
```

Comment:
I think we should move the logging until right before the command is actually run to avoid making it look like we started an exec session when actually we errored out before the command was run. It seems like the logging might look scary especially in the ACL error conditions where there are no log lines with the execID showing that the request failed. Operators looking for exec logs will be unable to determine if a session actually successfully executed a command or not.

Reply:
I can change logging to clarify whether request is successful or not - I think unsuccessful attempts by unauthorized users is just as a valuable to log as successful attempts.

Reply:
> unsuccessful attempts by unauthorized users is just as a valuable to log as successful attempts.

Will that be logged by the caller based on the error return? Seems like we can handle failures generically.

Reply:
Yeah, I refactored this to be `execImpl` that returns error/code which then `exec` logs.  `exec()` itself receives the connection and doesn't emit any error - which is the pattern for streaming rpc methods.